src/pytroll_watchers/dataspace_watcher.py

```diff
@@ -0,0 +1,198 @@
+        dataspace_auth: the authentication information, as a dictionary. It can be a dictionary with `username` and
+            `password` keys, or with `netrc_host` and optionaly `netrc_file` if the credentials are to be fetched with
+            netrc.
+        start_from: how far back in time to fetch the data the first time. This is helpful for the first iteration of
+            the generator, so that data from the past can be fetched, to fill a possible gap. Default to 0, meaning
+            nothing older than when the generator starts will be fetched. Same format accepted as polling_interval.
+        storage_options: The options to pass the S3Path instance, usually include ways to get credentials to the
+            copernicus object store, like `profile` from the .aws configuration files.
+
+    Yields:
+        Tuples of UPath (s3) and metadata.
+
+    """
+    with suppress(TypeError):
+        polling_interval = datetime.timedelta(**polling_interval)
+    with suppress(TypeError):
+        start_from = datetime.timedelta(**start_from)
+    if start_from is None:
+        start_from = datetime.timedelta(0)
+
+    last_pub_date = datetime.datetime.now(datetime.timezone.utc) - start_from
+
+    for next_check in run_every(polling_interval):
+        generator = generate_download_links_since(filter_string, dataspace_auth, last_pub_date, storage_options)
+        for s3path, metadata in generator:
+            last_pub_date = update_last_publication_date(last_pub_date, metadata)
+            yield s3path, metadata
+        logger.info("Finished polling.")
+        if next_check > datetime.datetime.now(datetime.timezone.utc):
+            logger.info(f"next iteration at {next_check}")
+
+
+def run_every(interval):
+    """Generator that ticks every `interval`.
+
+    Args:
+        interval: the timedelta object giving the amount of time to wait between ticks. An interval of 0 will just make
+        tick once, then return (and thus busy loops aren't allowed).
+
+    Yields:
+        The time of the next tick.
+    """
+    while True:
+        next_check = datetime.datetime.now(datetime.timezone.utc) + interval
+        yield next_check
+        to_wait = max(next_check.timestamp() - time.time(), 0)
+        time.sleep(to_wait)
+        if not interval:  # interval is 0
+           break
+
+
+def update_last_publication_date(last_publication_date, metadata):
+    """Update the last publication data based on the metadata."""
+    publication_date = _fromisoformat(metadata)
+    if publication_date > last_publication_date:
+        last_publication_date = publication_date
+    return last_publication_date
+
+
+def _fromisoformat(metadata):
+    try:
+        return datetime.datetime.fromisoformat(metadata["PublicationDate"])
+    except ValueError:
+        # for python 3.9
+        return datetime.datetime.strptime(metadata["PublicationDate"], "%Y-%m-%dT%H:%M:%S.%f%z")
```

Comment:
The try/execpt is not needed, pyproject.toml defines >= 3.10.

```suggestion
    return datetime.datetime.fromisoformat(metadata["PublicationDate"])
```

Reply:
Then I need to fix the ci.yaml file, thanks for spotting this

Reply:
Oh, my bad, the incompatibility is actually with 3.10 also, not just 3.9
```suggestion
    try:
        return datetime.datetime.fromisoformat(metadata["PublicationDate"])
    except ValueError:
        # for python 3.10
        return datetime.datetime.strptime(metadata["PublicationDate"], "%Y-%m-%dT%H:%M:%S.%f%z")
```